### ISSUE TITLE
H_res matrices don't learn with default parameters? Also a rabbit hole adventure on H_res properties and the effectiveness of them

### ISSUE THREAD

**Issue by pherber3** [tokenbender/mHC-manifold-constrained-hyper-connections#2]
**I just kept writing and this got super long so TL;DR**: Wanted to investigate Markov mixing in mHC's H_res matrices. Found that with default hyperparameters, gradients vanish and H_res stays at identity. Unclear if this matters since loss curves are similar before and after fixing tau, though different from the baseline no-mHC version. After fixing (higher tau), H_res learns but cumulative products converge toward uniform, meaning streams homogenize by late layers. The mHC paper's Figure 8 shows the same phenomenon but doesn't discuss it.  Maybe a residual-only ablation could isolate whether H_res actually contributes or if the branch is doing all the work.

I'm interested in studying the properties of the H_res matrices throughout training and the depth of the model, started off by running a small training example with:
```bash
python train.py config/train_fineweb10B_mhc.py max_iters=500 eval_interval=100 out_dir="out-mhc-baseline-test"
```
However, I saw that the H_res matrices stayed the exact same from training to the end at all layers of the model, i.e., staying as the identity matrix they were initialized as. 
```
transformer.h.5.hc_attn.H_res_logits:
  Logits:
tensor([[ 0., -8., -8., -8.],
        [-8.,  0., -8., -8.],
        [-8., -8.,  0., -8.],
        [-8., -8., -8.,  0.]])
  H_res (after Sinkhorn):
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
  Off-diagonal sample: -8.0000

transformer.h.5.hc_mlp.H_res_logits:
  Logits:
tensor([[ 0., -8., -8., -8.],
        [-8.,  0., -8., -8.],
        [-8., -8.,  0., -8.],
        [-8., -8., -8.,  0.]])
  H_res (after Sinkhorn):
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
```
Weird, not great since staying as the identity matrix means we essentially just have 4 parallel residual paths that don't interact with each other and that defeats the point of them. But, maybe the model wasn't deep enough for it to matter or something, so tried again with:
```bash
!python train.py config/train_fineweb10B_mhc_48l.py max_iters=500 eval_interval=100 out_dir="out-mhc-48l-old"
```
However, still ended up with the same thing overall, diagonal matrix all the way through (ended up cancelling the run ~3k steps):

<img width="1722" height="498" alt="Image" src="https://github.com/user-attachments/assets/1a274ecf-3bd5-4d8e-8c8b-120c320d27d1" />

Ran some tests and saw that the H_res matrices have a super low grad_norm in comparison to the other values:
```
H_res_logits:     grad_norm = 0.0000000061  (1e-9)
H_pre_logits:     grad_norm = 0.0025
branch.weight:    grad_norm = 1.8433
branch.bias:      grad_norm = 0.2500
```
I experimented with some additions mostly around the following and found something like this at least creates some change in the H_res matrices, though haven't optimized them or done a rigorous analysis. Honestly, my theory is that the only thing that really mattered was changing the tau value is the biggest thing since the gradients have to flow through the Sinkhorn projection and tau=0.05 was too sharp so the gradients were vanishing. 

```
sinkhorn_tau: 1.0 instead of 0.05
H_res_logits init: -2.0 off-diagonal, 0.0 diagonal (softer identity matrix initialization)
mhc_h_res_lr: Separate higher learning rate (0.1) for H_res_logits in the optimizer
```

> I think having -8 on the diagonal is probably fine and the tau value is the true issue, I tested with -8 initialization and it still learned, like I said didn't really try to optimize things. With tau = 0.05, we'd have something like exp(-8 / 0.05) = exp(-160) so maybe the Sinkhorn projection sees essentially zero probability on off-diagonal entries and gradients can't flow?  

That at least got me to something like this by the end of training the 48-layer for ~500 steps or so only this time (I'm GPU-poor):

<img width="1718" height="498" alt="Image" src="https://github.com/user-attachments/assets/7771c1e3-9327-4e4e-ab01-08e3f536ccf1" />

<img width="2051" height="238" alt="Image" src="https://github.com/user-attachments/assets/5f075950-ac85-4e15-8a6a-95ad3c5356ac" />

Doesn't really match their progression of H_res matrices from Figure 8 in the paper, but I'm not displaying the same layers and different model architecture, so whatever, cool they're at least changing now and doing *something*. Side note though our ~layer 30 displays do look similar though in terms of being near-uniform, which is what I was mostly interested in. But then looking at the W&B logs for both runs and a baseline no mHC run:

<img width="1470" height="309" alt="Image" src="https://github.com/user-attachments/assets/85bcf884-ef9e-4739-8306-eb6d0186f2a0" />

Overall performance is maybe marginally better with mHC, and also somehow even though the two H_res matrices are **different** for mHC runs, they have near-identical loss curves (green doesn't even show up they're so similar, off by like tenths to hundredths of a decimal point usually). This makes me wonder whether the H_res matrices here are just redundant for this model size maybe or what's going on exactly with them?

Some ideas:
- H_res learning dynamics don't affect final performance (at this scale)
- The branch term (H_pre/H_post) may be doing the heavy lifting

Anyways, originally all I wanted was that last cumulative product figure because, according to markov chain mixing, products of ergodic matrices converge to uniform. From an analysis on them I got:
```bash
Per-Layer Statistics (averaged):
  |λ₂| (second eigenvalue): 0.8905
  Spectral gap (1 - |λ₂|): 0.1095
```
The decay rate is |λ₂|^L, so after L H_res applications, stream diversity is roughly:

$$\text{diversity remaining} \approx |λ_2|^L = 0.8905^L$$

| H_res layers (L) | Diversity remaining |
|------------------|---------------------|
| 1 | 89% |
| 6 | 50% (half-life) |
| 12 | 25% |
| 24 | 6% |
| 48 | 0.4% |
| 96 | ≈0% |

**Half-life ≈ 6 H_res applications.** Every 6 H_res matrices, we lose half the stream differentiation.

For the 48-layer transformer (96 H_res modules counting attn+mlp), we'd expect streams to retain essentially 0% differentiation by the final layer. The cumulative product converges to (1/4)𝟙𝟙ᵀ which is just complete homogenization.

The first H_res product shows diagonal ~0.70 (streams distinct), but by product 96 all entries are ~0.25 (uniform matrix, streams identical).

I feel like the concerning read of this would be:

1. Early layers: H_res near identity, separate streams, but no mixing
2. Middle: H_res mixes
3. Late: Residual streams mix so thoroughly that they carry the same blended information and the multi-stream structure is useless since it's collapsed into effectively a single stream

**The paper's own Figure 8 shows similar convergence toward uniform at depth 60, so maybe this is expected behavior rather than a bug, but they didn't really discuss it at all so who knows**. I guess my next question is whether the branch term is what actually matters for mHC's effectiveness, making H_res somewhat redundant - which would explain why both tau values give similar loss despite very different H_res learning dynamics. Perhaps a residual-only ablation would isolate whether H_res actually contributes or if the branch is doing all the work, but I'm out of my free GPU credits for now rip. Maybe practically this doesn't matter, I don't really know enough to know whether it's something to be concerned about honestly. Apologies for the long post, initially just started with the hyperparameter stuff but maybe the rest is interesting as well lol.

**Comment by hjc18** [comment]
@pherber3 I agree. This implementation doesn't fully reproduce the paper. Maybe check out this repo: https://github.com/lucidrains/hyper-connections/blob/main/hyper_connections/mHCv2.py. It looks more consistent with the paper.
